### DISCUSSION TITLE
RE0085: closure cannot exist at runtime?

### DISCUSSION THREAD

**Discussion by DanBurton** [reach-sh/reach-lang#911] [Q&A] [answered]
Discord user DevFraccas asks:

Hello everyone! I'm curious if I'm able to create a smart contract, then at a later point in time go back in and update information as the Creator. In my particular case, I'd like to create a smart contract that contains an NFT for sale, then at a later point in time go back in as the Creator and bind a specified Buyer to the contract. 

This code successfully creates a contract in which only the specified Buyer can make the purchase. 

```js
'reach 0.1';
'use strict';


export const main = Reach.App(() => {
  const Seller =  Participant('Seller', {
    setSell: Fun([], Tuple(Token, UInt, UInt)),
    showContract: Fun([Contract], Null),
    setBuyer: Fun([], Address),
  });

  const Buyer =  Participant('Buyer', {
    getBuyInfo: Fun([Token, UInt], Null),
    showBoughtAsset: Fun([Token], Null),
  });

  init();

  // set NFT token id, nft amount, and price by Seller
  Seller.only(() => {
    const [ NFT, amtA, setPrice ] = declassify(interact.setSell());
  });
  Seller.publish(NFT, amtA, setPrice); // publish to blockchain
  commit();
  
  // Seller transfer NFTs to smart contract
  Seller.pay([ [amtA, NFT ]]);
  commit();

  // show seller creating contract the app id (smart contract)
  Seller.only(() => {
    interact.showContract(getContract());
  });

  // seller/creator sets the buyer (bind them to the contract)
  Seller.only(() => {
    const BuyerAddress = declassify(interact.setBuyer());
  });

  Seller.publish(BuyerAddress);

  // show the bound buyer the NFT asset ID and requested price
  Buyer.only(() => {
    check(Buyer == BuyerAddress, "Invalid Account");
    interact.getBuyInfo(NFT, setPrice);
  });

  commit();
  
  Buyer.pay(setPrice);

  transfer(balance()).to(Seller);
  transfer(balance(NFT), NFT).to(BuyerAddress);

  commit();

  Buyer.only(() => {
    interact.showBoughtAsset(NFT); 
  });

  exit();
});
```

The above example must set the Buyer during the contract creation. I'm now looking to create the contract and later bind the specified Buyer.

I'm playing around with the code below, but I'm not sure if this is the best way to achieve this. 

```js
'reach 0.1';
'use strict';


export const main = Reach.App(() => {
  const Creator = API('Creator', {
    setBuyer: Fun([], Address)
  });

  const Seller =  Participant('Seller', {
    setSell: Fun([], Tuple(Token, UInt, UInt)),
    showContract: Fun([Contract], Null),
  });

  const Buyer =  Participant('Buyer', {
    showBuyInfo: Fun([Token, UInt], Null),
    showBoughtAsset: Fun([Token], Null),
  });

  init();

  // set NFT token id, nft amount, and price by Seller
  Seller.only(() => {
    const [ NFT, amtA, setPrice ] = declassify(interact.setSell());
  });
  Seller.publish(NFT, amtA, setPrice); // publish to blockchain
  commit();
  
  // Seller transfer NFTs to smart contract
  Seller.pay([ [amtA, NFT ]]);
  commit();

  // show seller creating contract the app id (smart contract)
  Seller.only(() => {
    interact.showContract(getContract());
  });

  Seller.publish();
  commit();

  const BuyerAddress = call(Creator.setBuyer).assume(() => assume(true)); // not sure about this
  check(Seller == Creator, "Invalid Account");

  // show the bound buyer the NFT asset ID and requested price
  Buyer.only(() => {
    check(Buyer == BuyerAddress, "Invalid Account");
    interact.showBuyInfo(NFT, setPrice);
  });

  commit();
  
  check(Buyer == BuyerAddress, "Invalid Account");
  Buyer.pay(setPrice);

  transfer(balance()).to(Seller);
  transfer(balance(NFT), NFT).to(BuyerAddress);

  commit();

  Buyer.only(() => {
    interact.showBoughtAsset(NFT); 
  });

  exit();
});
```

I'm getting the following error in this example: 

```
reachc: error[RE0085]: Value cannot exist at runtime: [.api52.dom/59 : Tuple(), <closure> ]

  ./index.rsh:62:33:application

  62|   transfer(balance(NFT), NFT).to(BuyerAddress);
```

Is this the correct way to make an update in the contract? And if so, any thoughts as to why I'm getting this error? 

Error Info

RE0085
This error indicates that your program would contain a value at runtime which would not be allowed. This error usually stems from not fully applying a primitive function or using a value incorrectly, such as the participant interact interface of a Participant.

**Comment by DanBurton** [comment, marked as answer]
Notice that it's telling you the location of the error is the identifier `BuyerAddress` as it appears on this line

```
  transfer(balance(NFT), NFT).to(BuyerAddress);
```

You've defined `BuyerAddress` on this line:

```
  const BuyerAddress = call(Creator.setBuyer).assume(() => assume(true)); // not sure about this
```

The error message says that `BuyerAddress` has a `closure` in it which can't exist at runtime.
Check the docs for `call` to help understand why this is:


https://docs.reach.sh/rsh/step/#rsh_call

You should think of APIs from the caller's perspective. Instead of this:

```js
    setBuyer: Fun([], Address)
```

I think you want to do something like this:

```js
  setBuyer: Fun([Address], Null)
```

And then you write the `call` something like this:

```js
  const [[BuyerAddress], k] = call(Creator.setBuyer);
  k(null); // this is how you return the value expected by the caller of the API function; in this case just null
```

**Reply by DanBurton** [reply]
(`k` was the aforementioned closure, by the way.)

**Reply by Fraccas** [reply]
Awesome, the contract is now successfully compiling! 

I'm now getting this error on my frontend when trying to make the API call after the contract has been created by the Seller. 
```Error: Expected the DApp to be in state 3, but it was actually in state 2.```

I'm wondering if my frontend isn't set up properly. I'm attempting to complete the Seller steps, then make the API call 15 seconds after the contract has been created. 
```
const call = async (f) => {
      let res = undefined;
      try {
        res = await f();
      } catch (e) {
        res = [`err`, e]
      }
      console.log(`res`, res);
    };

    await call(() => backend.Creator_setBuyer(ctc, { // not sure about this
      setBuyer: () => {  return reach.formatAddress(buyer); } // bind buyer to contract
    }));
```

I've been following the API-Call example in the Reach repo. The backend.Creator_setBuyer(ctc, interact) was partially auto-generated for me as I started typing it out. 

I've also tried this with the same error message: 
```
    const Creator = ctc.a.Creator;
    await call(() => Creator.setBuyer(reach.formatAddress(buyer)));
```

Are you aware if there are any tutorials online for API calls? I feel like I don't fully understand how they work. I see the docs you posted regarding calls but it looks like it only covers how to use them in the RSH file.

**Reply by DanBurton** [reply]
Your second attempt is the correct calling convention.
Docs for how to call APIs are here:

https://docs.reach.sh/frontend/#calling-api-functions--ctcapis---ctca

`await ctc.a.Creator.setBuyer(someAddress)` is the correct way to call the `Creator.setBuyer` API function. (What you did was a fancy wrapper around basically this.)

**Reply by DanBurton** [reply]
The error you're getting says

```
Error: Expected the DApp to be in state 3, but it was actually in state 2.
```

This means that it is trying to call the API function prior to the Reach program arriving at the point when it is expected. My guess is that the second `Seller.publish()` on line 38 isn't actually completing before the Buyer calls the API function.

**Reply by Fraccas** [reply]
Ah yes, that makes sense. 

I'll continue to test to find out where it's going wrong. Thanks!

**Reply by Fraccas** [reply]
Thank you for all of your help @DanBurton. I was able to get this working. 

The final step is to ensure that the API call can only be made by the Creator/Seller. I ran some tests and it appears that anyone can set the value of BuyerAddress. 

```
// Seller transfer NFTs to smart contract
  Seller.pay([ [amtA, NFT ]]);
  commit();

  // show seller creating contract the app id (smart contract)
  Seller.only(() => {
    interact.showContract(getContract());
  });

  const [[BuyerAddress], k] = call(Creator.setBuyer); // issue: any account is allowed to set this value
  k(null); 

  // show the bound buyer the NFT asset ID and requested price
  Buyer.only(() => {
    check(Buyer == BuyerAddress, "Invalid Account");
    interact.showBuyInfo(NFT, setPrice);
  });

  commit();
```

Adding checks like below gives various different error messages during compilation. 
```   require(Seller == Creator, "Invalid Account"); ``` 

Are you aware of what is the best practice for this case?

**Reply by lorenzopolidori** [reply]
@Fraccas I wrote a similar contract [here](https://github.com/terragrids/dapp/blob/dev/blockchain/index.rsh). 

I use and Admin participant to put an NFT on the market, and an API for buyers to buy the NFT at fixed price. I am using a `parallelReduce` as I plan to extend the API for users to fetch more information about the NFT while it is on the market.
This way, users do not need to be always connected to the contract, not even the Admin.

A simple frontend test is [here](https://github.com/terragrids/dapp/blob/dev/blockchain/index.mjs).

cc - @DanBurton

**Reply by jeapostrophe** [reply]
@Fraccas You can check who calls an API by referring to `this` --- https://docs.reach.sh/rsh/consensus/#ref-programs-consensus-this --- so you'd do something like `check(this == Seller)` in the code for `Creator` calls that you want to constrain.

**Reply by Fraccas** [reply]
@lorenzopolidori thanks for the example. 

@jeapostrophe, would I need to convert my code to something like Lorenzo showed in their example? 

Lorenzo's example
```
const [done, buyer, paid] =
        parallelReduce([false, A, 0])
            .invariant(balance() == paid && balance(tok) == 1)
            .while(!done)
            .api(M.stop,
                (() => { assume(this == A); }),
                (() => 0),
                (k => {
                    const isAdmin = this == A;
                    k(isAdmin);
                    return [true, buyer, paid]
                }))
            .api(M.getToken,
                (k => {
                    k(tok);
                    return [false, buyer, paid]
                }))
            .api(M.buy,
                () => price,
                (k => {
                    k([this, price, tok]);
                    return [true, this, price + paid];
                }))
            .timeout(false);
```

My code uses call() rather than API(). 

```
  const [[BuyerAddress], k] = call(Creator.setBuyer);  //.assume(this == Seller);
  k(null); 
```

I'm not sure how to add the check(this == Seller) for the call(). Normally, I add these checks in a local step as shown below. 

```
  // show the bound buyer the NFT asset ID and requested price
  Buyer.only(() => {
    check(Buyer == BuyerAddress, "Invalid Account");
    interact.showBuyInfo(NFT, setPrice);
  });
```

**Reply by Fraccas** [reply]
Got it working! Thanks so much, @DanBurton  @jeapostrophe  @lorenzopolidori!! 

```
  const [[BuyerAddress], k] = call(Creator.setBuyer).assume((b) => { 
    check(this == Seller, "Invalid Account"); // ensure API caller is Seller
    check(this != b, "Invalid Account"); // ensure API caller isn't Buyer
  });
```
